scripts/s3mi

```diff
@@ -237,52 +248,37 @@ def main_raid(volume_name, *optional_args):
         return 1
     if len(optional_args) >= 1 and optional_args[0].startswith("/dev/"):
         return main_raid_nvme(volume_name, *optional_args)
-    else:
-        return main_raid_ebs(volume_name, *optional_args)
+    return main_raid_ebs(volume_name, *optional_args)
 
 
-def initiate_fetch(s3_bucket, s3_key, part, n, file_size, request_tokens, errors):
-    fetcher_subproc = None
-    watchdog = None
-    def kill():
-        if fetcher_subproc:
-            fetcher_subproc.terminate()
-    def note_error():
-        with errors[0]:
-            errors[1] += 1
-        tsprint("Fetching part '{}' failed.".format(part))
-    def wait_and_release():
-        try:
-            if fetcher_subproc and fetcher_subproc.wait() != 0:
-                note_error()
-        except:
-            note_error()
-            raise
-        finally:
-            request_tokens.release()
-            if watchdog:
-                watchdog.cancel()
+def append(part, baton, s3_bucket, s3_key, n, file_size, request_tokens):
     try:
         first = segment_start(n, file_size)
         last = segment_start(n + 1, file_size) - 1  # aws api wants inclusive bounds
         command_str = "aws s3api get-object --range bytes={rfrom}-{rto} --bucket {bucket} --key {key} {part}".format(
             rfrom=first, rto=last, bucket=s3_bucket, key=s3_key, part=part)
-        safe_remove(part)
-        os.mkfifo(part)
-        fetcher_subproc = subprocess.Popen(command_str.split(), stdout=DEVNULL)
-    except:
-        with errors[0]:
-            errors[1] += 1
+        for attempt in range(NUM_CHUNK_RETRIES):
+            try:
+                safe_remove(part)
+                os.mkfifo(part)
+                fetcher_subproc = subprocess.Popen(command_str.split(), stdout=DEVNULL)
+                watchdog = threading.Timer(TIMEOUT, fetcher_subproc.terminate)
+                try:
+                    watchdog.start()
+                    with open(part, 'rb') as pf:
+                        segment_bytes = pf.read()
+                    exitcode = fetcher_subproc.wait()
```

Comment:
wouldn't you want to wait before reading the dst file?

also, why not just call https://docs.python.org/2/library/subprocess.html#subprocess.Popen.poll in a loop instead of a timer?

Reply:
Interesting comment, @ttung -- in my testing I found out that if I kill the produced process for the pipe "pf", the consumer can hang on the pipe open system call here, line 268.

To work around that issue, I had to move this into a thread.   After doing that, I could then poll the subprocess in a loop from the main thread as you suggest.

Python 2.7.3+ has a timeout param in its subprocess wait() func, sadly we have to run this on machines with 2.7.12 at the moment.   So my hope is that in the near future I can just delete this watchdog business.

The sad thing is that after adding SIGTERM handler to these subprocesses, the code became a real mess --- stay tuned for the update here...  it will be disappointing.

Reply:
Note that the file here is a pipe, made with os.mkfifo a few lines above.